### PR TITLE
Cookie Builder

### PR DESCRIPTION
**Pull request by NicoNes**
A proposal to use builder pattern much more developer friendly than telescoping constructors in `Cookie` and `NewCookie` .

For example to set a `NewCookie` instance with a name and a `SameSite.LAX` attribute no longer need to do this:

`new NewCookie("name", null, null, null, Cookie.DEFAULT_VERSION, null, DEFAULT_MAX_AGE, null, false, false, SameSite.LAX);`

just do that:

`new NewCookie.Builder("name").sameSite(SameSite.LAX).build();`

### CODE REVIEW

**Comment by mkarg** [comment]
For backwards compatibility we must not *replace* the constructor, but we certainly can discuss adopting builders *in addition*. :-)

**Comment by mkarg** [comment]
I have reduced this MR to the draft state as it doesn't pass the CI build, and as it targets 3.1 but is backwards incompatible. Please either target 4.x *or* keep backwards compatibility, and please fix the build problems. When done, feel free to upgrade from draft to MR state again and I'll be glad to review it again. :-)

**Comment by NicoNes** [comment]
@mkarg Changes done to be keep backward compatibility.

Can I deprecate telescoping contructors ?

**Comment by mkarg** [comment]
I think we should discuss deprecating things and adding new things independently.

File: jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java
Comment (mkarg) [inline jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java:239]: Despite simply repeating the obvious, it might be beneficial to have some idea here what the builder actually is good for. There was a good reason why you invested your time to write it, so let the readers learn from this JavaDocs. :-)

File: jaxrs-api/src/test/java/jakarta/ws/rs/core/CookieTest.java
Comment (mkarg) [inline jaxrs-api/src/test/java/jakarta/ws/rs/core/CookieTest.java:49]: As we must guarantee backwards compatibility, we must not remove these tests.

Comment (NicoNes) [reply jaxrs-api/src/test/java/jakarta/ws/rs/core/CookieTest.java:49]: You are right I should have left existing tests unchanged and add new ones.

File: jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java
Comment (NicoNes) [reply jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java:239]: Well, builder pattern is a well known pattern now so I didn't think it worth explaining people why this pattern is better than unfriendly telescoping constructor (anti-)pattern.
By the way I got inspired by `OutboundSseEvent.Builder` javadoc so I thought it is acceptable.

Anyway I will changed it into something more like `jakarta.ws.rs.sse.SseEventSource.Builder`. :+1:

**Comment by NicoNes** [comment]
@mkarg changes done.
Feel free to correct my wording.
Note: javadoc on test `jakarta.ws.rs.core.NewCookieTest.testCtor` does not seems up to date.

**Comment by andymc12** [comment]
I'd be in favor of deprecating the constructors in this same PR since the builder pattern is effectively replacing them.

**Comment by mkarg** [comment]
> I'd be in favor of deprecating the constructors in this same PR since the builder pattern is effectively replacing them.

I have no strong feelings about that. So +-0.

**Comment by NicoNes** [comment]
@mkarg Do the changes about the JAVA doc look better/clearer to you ?

File: jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java
Comment (mkarg) [inline jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java:281]: What benefit brings `<T>`?

File: jaxrs-api/src/test/java/jakarta/ws/rs/core/CookieTest.java
Comment (mkarg) [inline jaxrs-api/src/test/java/jakarta/ws/rs/core/CookieTest.java:66]: The claims this test verifies are actually not rather obvious to detect due to the style of the test code. I do not enforce BDD, but it would be good if we could specify what the test actually wants to achieve, so other programmers can later easily understand and extend them.

File: jaxrs-api/src/test/java/jakarta/ws/rs/core/NewCookieTest.java
Comment (mkarg) [inline jaxrs-api/src/test/java/jakarta/ws/rs/core/NewCookieTest.java:82]: Do we really want to test two classes in the same single test?

**Review by mkarg — COMMENTED** [review]
No objections, but personally abstaining from vote as I do not miss this feature and love to see higher code quality for new APIs additions.

File: jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java
Comment (NicoNes) [reply jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java:253]: Well I fully agree with you. 
But you asked me to explain why the builder is good for here and you wanted me to let user discover it in the doc..so that's what I did. 
Sorry if I do not undertsand what you meant.

File: jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java
Comment (NicoNes) [reply jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java:277]: I can understand this.

File: jaxrs-api/src/test/java/jakarta/ws/rs/core/CookieTest.java
Comment (NicoNes) [reply jaxrs-api/src/test/java/jakarta/ws/rs/core/CookieTest.java:66]: Well again I agree with you but I did things the same way they are already inspired by existing test `CookieTest.testEquals()`.

File: jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java
Comment (mkarg) [reply jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java:253]: That's all fine. It's just the word "SHOULD" that bothers me in a specification project, as there are rules when to use SHOULD, MUST, etc.

File: jaxrs-api/src/test/java/jakarta/ws/rs/core/CookieTest.java
Comment (mkarg) [reply jaxrs-api/src/test/java/jakarta/ws/rs/core/CookieTest.java:66]: I know, and I am not proud of some things we did in the past, but let's try to make it even better together. :-)

File: jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java
Comment (NicoNes) [reply jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java:281]: Less code duplication in implementation of `AbstractCookieBuilder`

Comment (mkarg) [reply jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java:281]: I don't get it. Why not simply returning `AbstractCookieBuilder` (without `<T>`) in those methods that currently return `<T>`? Maybe I am blind. Can you give an example what you _actually_ achieve here?

File: jaxrs-api/src/test/java/jakarta/ws/rs/core/CookieTest.java
Comment (NicoNes) [reply jaxrs-api/src/test/java/jakarta/ws/rs/core/CookieTest.java:66]: +1 I agree with you.

I personnaly think that a JAVA specification should encourage all best design and code practices that people can use as a model for their development and thats why I made this proposal. 
Because IMO these telescoping constructors are no longer a good design/coding practice to encourage at all. Let me know if you agree.

You said that you prefer to see better code quality for new APIs addition and again I agree with you. 
Actually it's the APIs addition done by @chkal if I'm right, about samesite concept that leads me to this proposal. This new addition was simple and works fine but it introduces a 7 parameters constructor. Do you guys want to encourage this ?

Comment (mkarg) [reply jaxrs-api/src/test/java/jakarta/ws/rs/core/CookieTest.java:66]: Not unlimited. The sense of this specification is solely to define a common interface above already existing products, not to define a style of programming. We define just an API to use those products. JAX-RS is not a framework, not a product, not a language, and has _no particular_ style, neither a _preferred_ style. It must be long-living and cannot adjust to short-loving styles and fashions again and again. While we like builders today, we didn't in the past. Possibly we won't in the future. Certainly we try to _adjust_ builders where needed and useful. But it is **definitively not** our charter _to actively push_ programmers into _one particular_ programming style. That's completely up to the user of our API. Not vice versa.

**Review by mkarg — APPROVED** [review]

**Review by mkarg — COMMENTED** [review]
LGTM

File: jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java
Comment (NicoNes) [reply jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java:281]: This `AbstractCookieBuilder <T>` is used as a parent class of `AbstractNewCookieBuilder<T>`. 

Without the `T`, invocation of `NewCookie.Builder newCookieBuilder = new NewCookie.Builder("name").path("path")` returns an `AbstractCookieBuilder` type instead of a `NewCookie.Builder` type. So user can no longer invoke any of the specific methods (`comment`, `maxAge`...) define in `NewCookie.Builder`.

Comment (NicoNes) [reply jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java:281]: I found this link http://www.javabyexamples.com/lets-discuss-builder-pattern about it or this one https://medium.com/@hazraarka072/fluent-builder-and-powering-it-up-with-recursive-generics-in-java-483005a85fcd.
Let me know if clearer or not.

File: jaxrs-api/src/test/java/jakarta/ws/rs/core/CookieTest.java
Comment (NicoNes) [reply jaxrs-api/src/test/java/jakarta/ws/rs/core/CookieTest.java:66]: Well maybe "quality" was not the right world. IDK
But IMO whether or not an API is using telescoping contructors or builders is about code readability and maintainability.

By the way so that I know what is do you mean by "code quality" ?

**Comment by NicoNes** [comment]
Well, @mkarg altough you approved my proposal I think that together, we are not very comfortable with few things it contains.

And as I like to produce quality code just like you I guess, that's why I improved the tests. 
For sure more test could be written but for now I just reviewed the way current tests are written to improve readability and maintainability. Let me know what you think please.

About the part of the javadoc I added, I'm also not comfortable with it. As I said I just added it because you asked me too, but as you noticed it's more a global coding advice rather than a code documentation. 
More I don't think it is a good thing to tie the `Builder` javadoc to the contructor javadoc as I did. Because if we deprecate constructors and we later remove them in a future version, we will likely forget to update `Builder` javadoc. 
So if you agree I would like to remove this coding advice part from the  javadoc.

Another point about deprecating constructors. As I said already, I'm personnaly in favor of deprecating them in this PR just like @andymc12.
If all of you guys agree with that, I suggest to remove the 7 args constructor introduced by @chkal in `NewCookie` to support samesite. The idea is to be consistent not deprecating a contructor we just introduce in the same version (3.1).
It will also encourage people to use `Builder` instead. 
WDYT ?

Thanks

**Comment by mkarg** [comment]
> WDYT?

Sounds like a good plan to me. As your planned changes are rather big, I have reduced the status of this PR to *draft*.

> If all of you guys agree with that, I suggest to remove the 7 args constructor introduced by @chkal in NewCookie to support samesite.

Please do not *remove* constructors, even not this one, but just add deprecation info. I have no problem with introducing a new constructor in 3.1 which at the same time is already deprecated. By doing so we tell people clearly: *"Yes, we are fine with your old style of programming, but we think in future the constructor is not needed anymore thanks to the recently added builder."*

File: jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java
Comment (mkarg) [reply jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java:281]: While this seems to be a nice programming trick to finally get in Java what you like to achieve, it is not really improve the readability or easy comprehension of the code. This originally comes from (a) `<SELF>` being much more clear than `<T>` in the conext of that blog entries, and (b) making `NewCookie` a child of `Cookie` which IMHO was no smart design right from the start of JAX-RS. Anyways, you made your case very clear, thank you, so I am OK with it once you change `<T>` to `<SELF>` to better reflect that you applied that pattern, and if you like, introducing the `private SELF self` method as in that code example.

**Comment by mkarg** [comment]
@NicoNes This issue is on hold due to your last comments. What is the current status?

**Comment by NicoNes** [comment]
@mkarg All fixes has been done, the issue is ready for review.

File: jaxrs-api/src/test/java/jakarta/ws/rs/core/CookieTest.java
Comment (andymc12) [inline jaxrs-api/src/test/java/jakarta/ws/rs/core/CookieTest.java:99]:
```suggestion
		new Cookie(null, "value", "path", "domain");
```

extra whitespace here - PMD or CheckStyle may complain.

**Review by andymc12 — DISMISSED** [review]
LGTM - only comment about whitespace is in a test class, so it's probably not important unless it causes PMD or CheckStyle to complain.

File: jaxrs-api/src/test/java/jakarta/ws/rs/core/CookieTest.java
Comment (mkarg) [reply jaxrs-api/src/test/java/jakarta/ws/rs/core/CookieTest.java:99]: @NicoNes To finally merge your contribution to `master` I'd kindly like to ask the following of you:
* Please fix the above whitespace issue reported by @andymc12 (to prevent PMD / CheckStyle complaints).
* Rebase the whole PR ontop `master` to fix the CI fail on JDK 14+ (should be fixed by `master` already).

I will be happy to merge your change into `master` then.

Comment (mkarg) [reply jaxrs-api/src/test/java/jakarta/ws/rs/core/CookieTest.java:99]: @NicoNes Oh, and if you like, feel free to also mention it in https://github.com/eclipse-ee4j/jaxrs-api/blob/master/jaxrs-spec/src/main/asciidoc/chapters/appendix/_changes-since-3.0-release.adoc as part of the same PR. Thanks. :-)

Comment (NicoNes) [reply jaxrs-api/src/test/java/jakarta/ws/rs/core/CookieTest.java]: @mkarg and @andymc12 I made the fix and the rebase into `master` and I change the target branch of this PR to `master`.
I also mention the changes into the release document. Please feel free to correct my wording if necessary.

Thanks

**Comment by mkarg** [comment]
Kindly requesting reviews please, so we can merge this into `master`, i. e. have it contained in JAX-RS 3.1

**Review by chkal — APPROVED** [review]
I'm basically +1 for this.

I'm just not sure about the abstract build classes. I would prefer to simplify this aspect, even if this leads to some code duplication.

**Comment by NicoNes** [comment]
I had to rebase it due to JUNIT 5 migration so it dismissed your  approval @chkal

**Review by chkal — APPROVED** [review]

File: jaxrs-api/src/main/java/jakarta/ws/rs/core/NewCookie.java
Comment (andymc12) [inline jaxrs-api/src/main/java/jakarta/ws/rs/core/NewCookie.java:270]: Just noticed this, but since this constructor was added in 3.1, we should probably just delete it instead of deprecating it in the same release that we introduce it.

**Review by andymc12 — APPROVED** [review]
This looks good to me.  I'm torn on whether we should introduce a new constructor that is deprecated right out of the gate... If nobody else minds it, I won't complain. Thanks again for the work on this!

File: jaxrs-api/src/main/java/jakarta/ws/rs/core/NewCookie.java
Comment (NicoNes) [reply jaxrs-api/src/main/java/jakarta/ws/rs/core/NewCookie.java:270]: Well, that's is my opinion too but at the time @mkarg did not share it https://github.com/eclipse-ee4j/jaxrs-api/pull/895#issuecomment-679835532

Comment (andymc12) [reply jaxrs-api/src/main/java/jakarta/ws/rs/core/NewCookie.java:270]: :) This PR has been open for a while! I'd forgotten that conversation.  I'm fine with it as-is.  Thanks again!

Comment (mkarg) [reply jaxrs-api/src/main/java/jakarta/ws/rs/core/NewCookie.java:270]: We decided to just deprecate so the users have more time to adopt the new solution before we force them to do that.

**Comment by mkarg** [comment]
@spericas @jansupol @ronsigal Please vote for this PR. Thanks.

**Comment by mkarg** [comment]
@spericas @ronsigal @jansupol Please post your vote. Thanks.

**Comment by chkal** [comment]
Looks like there is a merge conflicts now. 

@NicoNes Could you perhaps fix this? That would be great! Thanks a lot!

**Comment by NicoNes** [comment]
@chkal merge conflict fixed.

**Review by andymc12 — DISMISSED** [review]

**Review by chkal — APPROVED** [review]

**Comment by mkarg** [comment]
I am hereby closing this PR as it did not reach the needed quorum.

**Comment by NicoNes** [comment]
@mkarg Well OK I'm not really surprised but I still keep on thinking that JAX-RS should propose another way to build cookie instead of a 11 arguments constructor.

**Comment by mkarg** [comment]
@NicoNes Nobody is *against* your proposal, there are just not enought people that like to *actively support* it. So maybe you can address some committers directly to get them into the boat? In the end, only a single vote was missing.

**Comment by jansupol** [comment]
I kinda like this

**Review by jansupol — DISMISSED** [review]

**Comment by jansupol** [comment]
@NicoNes Can you please update?

**Comment by NicoNes** [comment]
@jansupol Sure I'll do it !

**Comment by spericas** [comment]
No objections to using a builder pattern here

**Comment by NicoNes** [comment]
Rebase done but compilation failed in TCK module:

> StringBeanRuntimeDelegate is not abstract and does not override abstract method bootstrap(Class<? extends Application>,Configuration) in RuntimeDelegate

**Comment by spericas** [comment]
https://github.com/eclipse-ee4j/jaxrs-api/pull/1023

**Comment by mkarg** [comment]
Committers, those of you wanting this PR should take the time and review it a last time. The build is passing meanwhile, and if we want to have it in 3.1, then we should actively vote +1 **now**.  Just saying "I like it" is not enough, please really use the voting facitily of Github. Thanks. :-)

**Comment by mkarg** [comment]
If we add this, someone should add a TCK test for it ontop. As we recently discussed, we want to have new TCK tests for all new features.

**Comment by jansupol** [comment]
@mkarg There is no implementation required for this, everything is done in the API, what TCK test do you have in mind?

**Comment by mkarg** [comment]
> @mkarg There is no implementation required for this, everything is done in the API, what TCK test do you have in mind?

I am not a TCK expert, but I thought that the TCK also checks the existence of interface methods in the API JARs because vendors are allowed to privide their own .class files instead of using "ours". Maybe I am wrong here?

**Comment by jansupol** [comment]
@mkarg I believe that's what the signature tests are for. The signature tests make sure the API has all the methods it should have and no method is added.

**Review by jansupol — APPROVED** [review]

**Review by andymc12 — APPROVED** [review]

File: jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java
Comment (NicoNes) [reply jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java:279]: @mkarg asked me to rename it like that to improve undertstanding.
 I can change it if you all agree.

Comment (mkarg) [reply jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java:279]: Agreed.

Comment (mkarg) [reply jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java:279]: (while in fact, I like understandable code more than consistent code)

Comment (spericas) [reply jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java:279]: I think we should be consistent first and then if someone wants to go over this in a different PR (renaming type variables) we can do that too.

**Comment by spericas** [comment]
We may also need to relax the hidden field rule. Something like:

```
<module name="HiddenField" >
    <property name="ignoreSetter" value="true" />
    <property name="ignoreConstructorParameter" value="true" />
</module>
```

**Comment by NicoNes** [comment]
@spericas Done compilation is OK now

**Review by jansupol — APPROVED** [review]
LGTM. @NicoNes Thank you!

**Pull request merged**
